src/test/reconciliation_test.go

```diff
@@ -1227,6 +1243,91 @@ var _ = Describe("Reconciliation", func() {
 		})
 	})
 
+	When("the terminator has a node label selector", func() {
+		When("the label selector matches the target node", func() {
+			const labelName = "a-test-label"
+			const labelValue = "test-label-value"
+
+			BeforeEach(func() {
+				resizeCluster(3)
+
+				targetedNode, found := nodes[types.NamespacedName{Name: nodeNames[1]}]
+				Expect(found).To(BeTrue())
+
+				targetedNode.Labels = map[string]string{labelName: labelValue}
+
+				terminator, found := terminators[terminatorNamespaceName]
+				Expect(found).To(BeTrue())
+
+				terminator.Spec.MatchLabels = client.MatchingLabels{labelName: labelValue}
+
+				sqsQueues[queueURL] = append(sqsQueues[queueURL], &sqs.Message{
+					ReceiptHandle: aws.String("msg-1"),
+					Body: aws.String(fmt.Sprintf(`{
+						"source": "aws.ec2",
+						"detail-type": "EC2 Spot Instance Interruption Warning",
+						"version": "1",
+						"detail": {
+							"instance-id": "%s"
+						}
+					}`, instanceIDs[1])),
+				})
+			})
+
+			It("returns success and requeues the request with the reconciler's configured interval", func() {
+				Expect(result, err).To(HaveField("RequeueAfter", Equal(reconciler.RequeueInterval)))
+			})
+
+			It("cordons and drains only the targeted node", func() {
+				Expect(cordonedNodes).To(And(HaveKey(nodeNames[1]), HaveLen(1)))
+				Expect(drainedNodes).To(And(HaveKey(nodeNames[1]), HaveLen(1)))
+			})
+
+			It("deletes the message from the SQS queue", func() {
+				Expect(sqsQueues[queueURL]).To(BeEmpty())
+			})
+		})
+
+		When("the label selector does not match the target node", func() {
+			const labelName = "a-test-label"
+			const labelValue = "test-label-value"
+
+			BeforeEach(func() {
+				resizeCluster(3)
+
+				terminator, found := terminators[terminatorNamespaceName]
+				Expect(found).To(BeTrue())
+
+				terminator.Spec.MatchLabels = client.MatchingLabels{labelName: labelValue}
+
+				sqsQueues[queueURL] = append(sqsQueues[queueURL], &sqs.Message{
+					ReceiptHandle: aws.String("msg-1"),
+					Body: aws.String(fmt.Sprintf(`{
+						"source": "aws.ec2",
+						"detail-type": "EC2 Spot Instance Interruption Warning",
+						"version": "1",
+						"detail": {
+							"instance-id": "%s"
+						}
+					}`, instanceIDs[1])),
+				})
+			})
+
+			It("returns success and requeues the request with the reconciler's configured interval", func() {
+				Expect(result, err).To(HaveField("RequeueAfter", Equal(reconciler.RequeueInterval)))
+			})
+
+			It("does not cordon or drain any nodes", func() {
+				Expect(cordonedNodes).To(BeEmpty())
+				Expect(drainedNodes).To(BeEmpty())
+			})
+
+			It("does not delete the message from the SQS queue", func() {
```

Comment:
How long do messages stay on the SQS queue? It seems like there could be a failure mode where NTH spends all its time re-processing messages on the queue (because, e.g., the labels didn't match) and isn't able to respond to new messages in time for k8s to gracefully handle a node termination. That's probably uncommon, but it is basically what mechanical-fish described in https://github.com/aws/aws-node-termination-handler/issues/494.

Is that a concern with this way of handling node labels? It seems like the philosophy of the terminator is "I only respond to requests that match my parameters; anything else I will leave for someone else to handle". So it may not be possible to prevent that situation. Curious what your thoughts are.

Reply:
Queues have a [message retention period](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-configure-queue-parameters.html) setting. A [dead-letter queue](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-dead-letter-queues.html) can also be configured.

Leaving the message in the queue gives another Terminator instance a chance to match and handle it.